lib/api_client/src/client.rs

```diff
@@ -0,0 +1,39 @@
+use anyhow::Context;
+use graphql_client::GraphQLQuery;
+use url::Url;
+
+pub type QueryResult<T> =
+    anyhow::Result<graphql_client::Response<<T as GraphQLQuery>::ResponseData>>;
+
+pub struct Client {
+    url: Url,
+}
+
+impl Client {
+    pub fn new(url: Url) -> Self {
+        Self { url }
+    }
+
+    /// Issue a GraphQL query using Reqwest, serializing the response to the associated
+    /// GraphQL type for the given `request_body`
+    pub async fn query<T: GraphQLQuery>(
+        &self,
+        request_body: &graphql_client::QueryBody<T::Variables>,
+    ) -> QueryResult<T> {
+        let client = reqwest::Client::new();
```

Comment:
This could perhaps be stored against the `Client` instance directly, but I'm not sure if there's any interior mutability in play between requests, so I just left it as a new instance for each query.

Reply:
Yes, we should definitely do this. All the interesting calls are non-mut. Furthermore, we should definitely allow passing a custom `reqwest` instance - so that users can specify thier own preferences for the connection.

I suggest adding a builder (using https://docs.rs/derive_builder/0.9.0/derive_builder/) and accepting both `url` and `reqwest::Client`.

Reply:
I've never used a builder pattern in Rust- this looks interesting. To avoid blocking this PR, I'll go ahead and open a separate issue to learn a bit more about this, and add it later. But thanks for bringing that to my attention.

Reply:
https://github.com/timberio/vector/issues/4538

Reply:
`derive_builder` is a syntax sugar. The idea there is to, rather than taking `new(url: _, client: reqwest::Client)` which requires passing the `client`, or `new(url: _, client: Option<reqwest::Client>)` which can be though of "pass request, or we'll use `reqwest::Client::default()`, but in a more elegant way.

With builder (and `derive_builder` simply to type less code :smile:) it's possible to make `url` be mandatory, and `reqwest::Client` optional, and default-initialized if none passed.

Reply:
Yeah, gotcha. Makes sense. I've wanted to reach for a pattern like that a couple of times, but didn't know there was an idiomatic solution for it. I'll use that in the next round. My immediate use-case for `Url` is to unblock the `vector top` work, so I'll defer this until I've had chance to read the above link and make sure I'm doing it properly -- but it sounds ideal.